src/transactions/cpfp.rs

```diff
@@ -0,0 +1,181 @@
+use crate::{
+    error::*,
+    transactions::{utils, CpfpableTransaction, RevaultTransaction, CPFP_MIN_CHANGE},
+    txins::*,
+    txouts::*,
+};
+
+use miniscript::bitcoin::{
+    consensus::encode::Decodable,
+    util::psbt::{Input as PsbtIn, PartiallySignedTransaction as Psbt},
+    Amount, Script, SigHashType, Transaction, TxIn, TxOut,
+};
+
+#[cfg(feature = "use-serde")]
+use {
+    serde::de::{self, Deserialize, Deserializer},
+    serde::ser::{Serialize, Serializer},
+};
+
+impl_revault_transaction!(
+    CpfpTransaction,
+    doc = "The transaction feebumping either an unvault or a spend"
+);
+impl CpfpTransaction {
+    /// Create a new cpfp transaction from the transaction that needs cpfp, its fees and weight;
+    /// the required feerate (expressed in sats/WU); a list of CpfpTxIn that can be used along
+    /// with their child number, the cpfp descriptor and the secp curve.
+    pub(crate) fn from_tx(
+        to_be_cpfped: &impl CpfpableTransaction,
+        feerate: f64,
+        mut listunspent: Vec<CpfpTxIn>,
+    ) -> Result<CpfpTransaction, TransactionCreationError> {
+        assert!(to_be_cpfped.is_finalized());
+        // Checking that the feerate requested is higher than the previous
+        // tx feerate (otherwise why cpfping in the first place?)
+        let to_be_cpfped_weight = to_be_cpfped.clone().into_tx().get_weight();
+        let to_be_cpfped_fees = Amount::from_sat(to_be_cpfped.fees());
+
+        if to_be_cpfped_fees.as_sat() as f64 / to_be_cpfped_weight as f64 >= feerate {
```

Comment:
alarms buzzing in my head reading divisions plus floats. How about:
```rust
let current_fees = Amount::from_sat(to_be_cpfped_weight * feerate);
if to_be_cpfped_fees >  current_fees {}
```

Also, just taking a feerate *increase* as argument of this function would implicitly get you rid of this error case. It also makes sense from an API perspective IMO.

Reply:
> It also makes sense from an API perspective IMO.

I don't think it does. The client does know the package feerate he wants to target, but he doesn't really know the feerate increase. He would need to calculate the tx feerate and then calculate the increase, but this would mean calculating the initial tx feerate twice (once on client side, once here, as we need it anyways)